File: python/mirage/runtime/python/monty/stat.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.runtime.python.monty.binding import StatResult
from mirage.runtime.types import VFSStat


def stat_result(st: VFSStat) -> Any:
    # 0 is the door's spelling of "no stamp", and monty reads a 0.0
    # as epoch zero rather than substituting the host clock, so an
    # unknown mtime stays unknown instead of becoming now.
    mtime = st.mtime_ns / 1_000_000_000
    if st.is_dir:
        return StatResult.dir_stat(mode=st.mode, mtime=mtime)
    return StatResult.file_stat(size=st.size, mode=st.mode, mtime=mtime)

File: typescript/packages/core/src/runtime/js/quickjs/stat.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { QuickJSAsyncContext, QuickJSHandle } from 'quickjs-emscripten'
import type { RuntimeVFS, VFSStat } from '../../vfs.ts'
import { WASI } from './wasi.ts'
import { wasiErrno } from './errors.ts'

export async function stat(
  ctx: QuickJSAsyncContext,
  vfs: RuntimeVFS | null,
  pathH: QuickJSHandle,
): Promise<QuickJSHandle> {
  const path = ctx.getString(pathH)
  let st: VFSStat | null = null
  let errno = 0
  if (!vfs?.mountOf(path)) {
    errno = WASI.ENOENT
  } else {
    try {
      st = await vfs.stat(path)
    } catch (err) {
      errno = wasiErrno(err)
    }
  }
  const tuple = ctx.newArray()
  if (st === null) {
    ctx.setProp(tuple, 0, ctx.null)
  } else {
    const obj = ctx.newObject()
    const setNum = (key: string, value: number): void => {
      const h = ctx.newNumber(value)
      ctx.setProp(obj, key, h)
      h.dispose()
    }
    setNum('dev', 0)
    setNum('ino', 0)
    setNum('mode', st.mode)
    setNum('nlink', 1)
    setNum('uid', 0)
    setNum('gid', 0)
    setNum('rdev', st.rdev ?? 0)
    setNum('size', st.size)
    setNum('blocks', Math.ceil(st.size / 512))
    setNum('atime', st.mtimeMs)
    setNum('mtime', st.mtimeMs)
    setNum('ctime', st.mtimeMs)
    ctx.setProp(tuple, 0, obj)
    obj.dispose()
  }
  const errH = ctx.newNumber(errno)
  ctx.setProp(tuple, 1, errH)
  errH.dispose()
  return tuple
}
